### PR TITLE
Implemented a ton of UGC features.

### PR DESCRIPTION
**Pull request by bubby932**
* Subroom link verification
* Significantly more documentation on types.
* Get instance info
* Permission update notifications
* Added role deletion
* Added role viewing
* Added user role updates
* Added `/api/rooms/all-permissions`
* Fixed an issue where the notifications menu would be broken by a certain notification format.
* Granted developers all permissions in every room, serverside instead of just clientside.

### CODE REVIEW

File: routers/matchmaking.js [hunk 1]
```diff
@@ -48,49 +48,126 @@ router.get("/:room_id/:subroom_id/all-instances", middleware.authenticateDevelop
     }
 });
 
+router.get("/my-instance-info", middleware.authenticateToken, async (req, res) => {
```
Comment (bubby932) [inline routers/matchmaking.js:51]: New endpoint used for determining what kind of instance to create when traversing a Subroom Link.

File: routers/matchmaking.js [hunk 2]
```diff
@@ -48,49 +48,126 @@ router.get("/:room_id/:subroom_id/all-instances", middleware.authenticateDevelop
...
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal error occurred and we couldn't serve your request."
+        });
+        throw ex;
+    }
+});
 
-// yes this is literally just an enum
+/**
```
Comment (bubby932) [inline routers/matchmaking.js:84]: Type documentation

File: routers/matchmaking.js [hunk 3]
```diff
@@ -48,49 +48,126 @@ router.get("/:room_id/:subroom_id/all-instances", middleware.authenticateDevelop
...
-// By design, when the API goes down, all users are removed from the game due to their websockets disconnecting.
-// This prevents failed API calls and cheating by faking a websocket connection.
-// HOWEVER - This also means that instances created by Photon will slowly be removed from the room list when all users are disconnected.
-// This means that the RoomSessions dictionary will be OUT OF DATE if it persists.
-// Therefore, it is in memory. This allows for both faster read/writes, better class implementation, more intellisense options,
-// and preventing the aforementioned persistence issues.
-
-// "Well, just make it so Photon Instances never expire, and you can persist as long as you like"
-// THIS IS A WASTE OF RESOURCES. It may also incur additional fees from Photon.
+/**
```
Comment (bubby932) [inline routers/matchmaking.js:94]: A significant amount of incredibly useful documentation with IDE support using JSDoc.

File: routers/matchmaking.js [hunk 4]
```diff
@@ -171,19 +248,25 @@ class RoomSession {
         this.JoinCode = `CVR_ROOM.${this.RoomId}.INSTANCE.${this.InstanceId}`;
 
         // global instance handling
-        this.GlobalInstanceId = GlobalInstanceId === null ? uuid.v1() : GlobalInstanceId;
+        this.GlobalInstanceId = GlobalInstanceId == null ? uuid.v1() : GlobalInstanceId;
 
         if(typeof GlobalRoomInstances[RoomId] != 'object') GlobalRoomInstances[RoomId] = {};
 
+        var x = {};
```
Comment (bubby932) [inline routers/matchmaking.js:255]: Fixed some matchmaking logic errors.

File: routers/matchmaking.js [hunk 5]
```diff
@@ -171,19 +248,25 @@ class RoomSession {
...
-            GlobalRoomInstances[RoomId][this.GlobalInstanceId] = {SubroomId: this.InstanceId};
+            GlobalRoomInstances[RoomId][this.GlobalInstanceId] = x;
         else 
             GlobalRoomInstances[RoomId][this.GlobalInstanceId][SubroomId] = this.InstanceId;
 
         this.BeginEventLoop();
     }
 }
 
+/**
```
Comment (bubby932) [inline routers/matchmaking.js:267]: More type docs

File: routers/rooms.js [hunk 6]
```diff
@@ -207,7 +207,7 @@ router.get("/search", authenticateToken_optional, async (req, res) => {
     }
 });
 
-router.put('/room/:id/subrooms/:subroom_id/versions/new', authenticateToken, canViewRoom, async (req, res) => {
+router.put('/room/:id/subrooms/:subroom_id/versions/new', authenticateToken, requiresRoomPermission("createVersions"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:210]: Standardization of permissions

File: routers/rooms.js [hunk 7]
```diff
@@ -273,11 +273,6 @@ router.put('/room/:id/subrooms/:subroom_id/versions/new', authenticateToken, can
             "message": "The `collaborators` parameter of your version metadata is not specified or is invalid."
         });
 
-        if(!(await hasPermission(req.user.id, id, 'createVersions'))) return res.status(403).json({
```
Comment (bubby932) [inline routers/rooms.js:276]: See previous

File: routers/rooms.js [hunk 8]
```diff
@@ -308,7 +303,7 @@ router.put('/room/:id/subrooms/:subroom_id/versions/new', authenticateToken, can
         throw ex;
     }
 });
-router.post('/room/:id/subrooms/:subroom_id/versions/:version_id/associate-data', authenticateToken, canViewRoom, async (req, res) => {
+router.post('/room/:id/subrooms/:subroom_id/versions/:version_id/associate-data', authenticateToken, requiresRoomPermission("createVersions"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:306]: More standardization

File: routers/rooms.js [hunk 9]
```diff
@@ -384,7 +374,7 @@ router.post('/room/:id/subrooms/:subroom_id/versions/:version_id/associate-data'
         throw ex;
     }
 });
-router.post('/room/:id/subrooms/:subroom_id/versions/public', authenticateToken, canViewRoom, async (req, res) => {
+router.post('/room/:id/subrooms/:subroom_id/versions/public', authenticateToken, requiresRoomPermission("setPublicVersion"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:377]: And more

File: routers/rooms.js [hunk 10]
```diff
@@ -493,7 +478,7 @@ router.post('/room/:id/tags', authenticateToken, requiresRoomPermission("manageT
     }
 });
 
-router.post('/room/:id/content_flags', authenticateToken, requiresRoomPermission("manageTags"), async (req, res) => {
+router.post('/room/:id/content_flags', authenticateToken, requiresRoomPermission("manageContentFlags"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:481]: And more

File: routers/rooms.js [hunk 11]
```diff
@@ -626,12 +611,12 @@ router.post('/room/:id/moderation-suspend', authenticateDeveloperToken, async (r
                 $push: {
                     "notifications": {
                         template: "room_suspension_notice",
-                        data: {
+                        parameters: {
```
Comment (bubby932) [inline routers/rooms.js:614]: Fixes a critical bug in the Notifications menu.

File: routers/rooms.js [hunk 12]
```diff
@@ -626,12 +611,12 @@ router.post('/room/:id/moderation-suspend', authenticateDeveloperToken, async (r
...
                             "headerText": "<smallcaps><color=red>Urgent Moderation Notice",
                             "bodyText": `We regret to inform you that your room <noparse>"${roomname}"</noparse> has been <color=yellow>suspended</color> by the Compensation Social moderation team.
 
-                            For more information, please see
-                            <color=#FF5566>https://compensationvr.tk/about/suspension</color>`
+For more information, please see
```
Comment (bubby932) [inline routers/rooms.js:618]: Fixes TMP tab handling.

File: routers/rooms.js [hunk 13]
```diff
@@ -752,14 +737,14 @@ router.post("/room/:id/moderation-terminate", authenticateDeveloperToken, async
...
-                        data: {
+                        parameters: {
                             "headerText": "<smallcaps><color=red>Urgent Moderation Notice",
                             "bodyText": `We regret to inform you that your room <noparse>"${roomname}"</noparse> has been <color=#FF5566>Terminated</color> by the Compensation Social moderation team.
 
-                            For more information, please see
-                            <color=#FF5566>https://compensationvr.tk/about/terminated</color>
-                            
-                            You can appeal this decision on our Discord, the link is available at the page above.`
+For more information, please see
```
Comment (bubby932) [inline routers/rooms.js:744]: More tab fixes

File: routers/rooms.js [hunk 14]
```diff
@@ -752,14 +737,14 @@ router.post("/room/:id/moderation-terminate", authenticateDeveloperToken, async
                 $push: {
                     "notifications": {
                         template: "room_termination_notice",
-                        data: {
+                        parameters: {
```
Comment (bubby932) [inline routers/rooms.js:740]: Another critical bug fix.

File: routers/rooms.js [hunk 15]
```diff
@@ -1227,7 +1203,37 @@ router.put('/room/:id/roles/new', authenticateToken, canViewRoom, async (req, re
     }
 });
 
-router.put("/room/:id/roles/:role_name/update", authenticateToken, requiresRoomPermission("managePermissions"), async (req, res) => {
+router.get('/room/:id/permissions', authenticateToken, requiresRoomPermission("viewPermissions"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1206]: Endpoint used for fetching the roles and user permissions of a room.

File: routers/rooms.js [hunk 16]
```diff
@@ -1271,7 +1277,7 @@ router.put("/room/:id/roles/:role_name/update", authenticateToken, requiresRoomP
         for (const key in permissions) {
             if (Object.hasOwnProperty.call(permissions, key)) {
                 const element = permissions[key];
-                if (!req.userRoomPermissions[key]) return res.status(400).json({
+                if (!req.userRoomPermissions[key] && req.userRoomRole != "owner" && !req.user.developer) return res.status(400).json({
```
Comment (bubby932) [inline routers/rooms.js:1280]: Developers should have all permissions here.

File: routers/rooms.js [hunk 17]
```diff
@@ -1302,10 +1308,18 @@ router.put("/room/:id/roles/:role_name/update", authenticateToken, requiresRoomP
             }
         );
 
+        const keys = Object.keys(room.userPermissions);
+
         res.status(200).json({
             code: "success",
             message: "The operation was successful."
         });
+
```
Comment (bubby932) [inline routers/rooms.js:1317]: permission update message

File: routers/rooms.js [hunk 18]
```diff
@@ -1315,6 +1329,146 @@ router.put("/room/:id/roles/:role_name/update", authenticateToken, requiresRoomP
     }
 });
 
+router.post("/room/:id/roles/:role_name/delete", authenticateToken, requiresRoomPermission("managePermissions"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1332]: Role deletion

File: routers/rooms.js [hunk 19]
```diff
@@ -1315,6 +1329,146 @@ router.put("/room/:id/roles/:role_name/update", authenticateToken, requiresRoomP
...
+        });
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal server error occurred and we couldn't serve your request."
+        });
+    }
+});
+
+router.post("/room/:id/user/:user_id/set-role/:role_name", authenticateToken, requiresRoomPermission("managePermissions"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1402]: User permission management

File: routers/rooms.js [hunk 20]
```diff
@@ -1315,6 +1329,146 @@ router.put("/room/:id/roles/:role_name/update", authenticateToken, requiresRoomP
...
+            {
+                _id: {
+                    $eq: id,
+                    $exists: true
+                }
+            },
+            update
+        );
+
+        require('./ws/WebSocketServerV2').ws_connected_clients[user_id]?.socket.emit('permission-update', id);
```
Comment (bubby932) [inline routers/rooms.js:1457]: Another use of the permission-update websocket packet.

File: routers/rooms.js [hunk 21]
```diff
@@ -1371,6 +1525,75 @@ router.post("/room/:id/cover-image/set/:image_id", authenticateToken, requiresRo
     }
 });
 
+router.get("/room/:id/verify-subroom-link/:to", authenticateToken, canViewRoom, async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1528]: Subroom link verification.

File: routers/rooms.js [hunk 22]
```diff
@@ -1371,6 +1525,75 @@ router.post("/room/:id/cover-image/set/:image_id", authenticateToken, requiresRo
...
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal server error occurred while processing your request.",
+            valid: false
+        });
+        throw ex;
+    }
+});
+
+router.get("/all-permissions", async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1564]: All permissions endpoint, used for listing off available permissions to set in the role management menu.

File: routers/rooms.js [hunk 23]
```diff
@@ -1434,7 +1657,7 @@ function requiresRoomPermission(permission) {
     
         const role = Object.keys(userPermissions).includes(req.user.id) ? userPermissions[req.user.id] : "everyone";
     
-        if(!req.user.developer && rolePermissions[role][permission]) {
+        if(!req.user.developer && !rolePermissions[role][permission] && role != "owner") {
```
Comment (bubby932) [inline routers/rooms.js:1660]: Owners of the room should be able to assign any permission or role.

File: routers/rooms.js [hunk 24]
```diff
@@ -1448,6 +1671,9 @@ function requiresRoomPermission(permission) {
     };
 }
 
+
```
Comment (bubby932) [inline routers/rooms.js:1674]: Fixes ESLint screaming, this function can likely be removed soon.

File: routers/ws/WebSocketServerV2.js [hunk 25]
```diff
@@ -5,12 +5,45 @@ const WebSocket = require('ws');
 const { MatchmakingModes } = require('../matchmaking');
 const { WebSocketV2_MessageTemplate } = require("../../index");
 const { auditLog } = require('../../helpers');
+
```
Comment (bubby932) [inline routers/ws/WebSocketServerV2.js:8]: More wonderful type documentation.

File: routers/ws/WebSocketServerV2.js [hunk 26]
```diff
@@ -5,12 +5,45 @@ const WebSocket = require('ws');
...
+ */
+
+/**
+ * @type {Object.<string, Connection>}
+ */
 var ws_connected_clients = {};
+
 exports.ws_connected_clients = ws_connected_clients;
 
+/**
```
Comment (bubby932) [inline routers/ws/WebSocketServerV2.js:27]: Another typedef used for preventing logic errors in advance.

File: routers/ws/WebSocketServerV2.js [hunk 27]
```diff
@@ -664,6 +697,32 @@ WebSocketServerV2.on('connection', (Socket) => {
         };
         Socket.send(JSON.stringify(send, null, 5));
     });
+    Socket.on('permission-update', async (roomId) => {
```
Comment (bubby932) [inline routers/ws/WebSocketServerV2.js:700]: Permission update websocket message handler.

**Review by bubby932 — COMMENTED** [review]
Explained all changes fairly thoroughly. Ready for review.

**Review by Bobrobot1 — APPROVED** [review]

**Pull request merged**